server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -345,6 +346,9 @@ public void applyClusterState(ClusterChangedEvent event) {
                 }
             }
             repositories = Collections.unmodifiableMap(builder);
+            for (Repository repo : repositories.values()) {
+                repo.updateState(state);
```

Comment:
I think we need to catch a potential RepositoryException here, log it and continue to update the others repositories.

Reply:
I wonder, looking at the code above we aren't defensive around `closeRepository` either (though that might be called in a loop as well) but catch (which has a TODO about it) on `createRepository`.
Maybe we should rather wrap this in a 

```java
try {
    repo.updateState(state);
} catch (Exception e) {
   assert false;
   throw e;
}
```

and do the same for `closeRepository`? Could do it in a follow up and remove that todo while we're at it :)

Reply:
My main contention to just catch and log here would be that once the state update is important for the proper functioning of blob store repositories, then what would an uncaught exception even mean? (imo it would mean that the repo can't be used for writes any longer ... but that would be something the repo would have to set in its internal state when handling exceptions)

Reply:
The bigger issue I think is to get rid of the condition:
```
 if ((oldMetaData == null && newMetaData == null) || (oldMetaData != null && oldMetaData.equals(newMetaData))) {
```

which currently compares the current with the previous cluster state. If anything went wrong applying the previous cluster state, we will not update the repo again.